R/text_tools.R

```diff
@@ -119,58 +81,269 @@ extract_location <- function(v) {
+  if (inherits(v, "data.frame")) {
+    if (!"doc_id" %in% names(v)) {
+      stop("Please declare a text vector or an annotated data frame.")
+    }
+    if ("sentence" %in% names(v)) {
+      v <- v[["sentence"]]
+    } else if ("token_id" %in% names(v)) {
+      v <- dplyr::group_by(doc_id) %>%
+        dplyr::summarise(text = paste(token, collapse = " ")) %>%
+        dplyr::select(text)
+    }
+  }
+  if (invert == TRUE && ignore.case == FALSE) {
+    t <- lapply(v, function(x) grep(match, x, value = TRUE,
+                                       ignore.case = FALSE, invert = TRUE))
+  } else if (invert == TRUE && ignore.case == TRUE) {
+    t <- lapply(v, function(x) grep(match, x, value = TRUE,
+                                       ignore.case = TRUE, invert = TRUE))
+  } else if (invert == FALSE && ignore.case == FALSE) {
+    t <- lapply(v, function(x) grep(match, x, value = TRUE, ignore.case = FALSE))
+  } else {
+    t <- lapply(v, function(x) grep(match, x, value = TRUE, ignore.case = TRUE))
+  }
+  t
+}
+
+#' Extract context for string matches
+#'
+#' A function for getting string matches and the context in which they occur.
+#' @param v Text vector or annotated data frame.
+#' @param match Character string to be matched.
+#' For multiple strings, please use "|" as a separator.
+#' @param level At which text level do you want matches to be returned?
+#' Defaults to "sentences".
+#' Options are sentences, words, and paragraph.
+#' @param n Number of sentences or words matched before and after string match.
+#' Defaults to 1.
+#' That is, one word or one sentence before, and after, string match.
+#' For paragraphs, n is always set to one.
+#' @importFrom stringr str_detect str_extract_all
+#' @importFrom dplyr group_by summarise select %>%
+#' @examples
+#' \donttest{
+#' extract_context(match = "war|weapons of mass destruction|conflict|NATO|peace",
+#'                 v = US_News_Conferences_1960_1980$text[100],
+#'                 level = "sentences", n = 2)
+#' }
+#' @return A list of string matches an their context
+#' @export
+extract_context <- function(match, v, level = "sentences", n = 1) {
+  doc_id <- sentence <- text <- token <- NULL
+  if (inherits(v, "data.frame")) {
+    if (!"doc_id" %in% names(v)) {
+      stop("Please declare a text vector or an annotated data frame.")
+    }
+    if ("sentence" %in% names(v)) {
+      v <- dplyr::group_by(doc_id) %>%
+        dplyr::summarise(text = paste(sentence, collapse = " ")) %>%
+        dplyr::select(text)
+    } else if ("token_id" %in% names(v)) {
+      v <- dplyr::group_by(doc_id) %>%
+        dplyr::summarise(text = paste(token, collapse = " ")) %>%
+        dplyr::select(text)
+    }
+  }
+  if (level == "sentences") {
+    s <- stringr::str_extract_all(v, paste0("([^.]+\\.){0,", n, "}[^.]+(", match, ").*?\\.([^.]+\\.){0,", n, "}"))
+  } else if (level == "words") {
+    s <- stringr::str_extract_all(v, paste0("([^\\s]+\\s+){", n,"}", match, "(\\s+[^\\s]+){", n, "}"))
```

Comment:
Do you want to break this up/add in-line comments so that future Henrique/contributor knows what all the regex is referring to?